app/helpers/application_helper.rb

```diff
@@ -320,9 +320,20 @@ def link_to_chart(name, values)
   def deployed_or_running_list(stages, reference)
     html = "".html_safe
     stages.each do |stage|
-      next unless deploy = stage.deployed_or_running_deploy
-      next unless deploy.references?(reference)
-      label = (deploy.active? ? "label-warning" : "label-success")
+      # The first deploy is the most recent one.
+      next unless deploy = stage.deploys.where(reference: reference).first
```

Comment:
Instead of `next` and doing n where queries, would it be better to do something like:
```ruby
deploys_hash = Deploy.joins(:stage).where(stage_id: stages.pluck(:id)).where(reference: 'master').group_by(&:stage_id)
stages.each do |stage|
  next unless deploy = deploys_hash[stage.id]&.first
  ...
end
```
?
```
samson(dev)> puts Benchmark.measure { deployed_or_running_list(stages, reference)  }
  Deploy Load (0.6ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 4 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  Deploy Load (0.5ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 2 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  Deploy Load (0.4ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 3 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  Deploy Load (0.4ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 1 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  Deploy Load (0.4ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 5 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  Deploy Load (0.3ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 6 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  Deploy Load (0.4ms)  SELECT  `deploys`.* FROM `deploys` WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` = 7 AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC LIMIT 1
  0.005582   0.000476   0.006058 (  0.008085)
=> nil


samson(dev)> puts Benchmark.measure { deployed_or_running_list2(stages, reference)  }
  Deploy Load (1.0ms)  SELECT `deploys`.* FROM `deploys` INNER JOIN `stages` ON `stages`.`id` = `deploys`.`stage_id` AND `stages`.`deleted_at` IS NULL WHERE `deploys`.`deleted_at` IS NULL AND `deploys`.`stage_id` IN (4, 2, 3, 1, 5, 6, 7) AND `deploys`.`reference` = 'master' ORDER BY `deploys`.`id` DESC
  0.002177   0.000184   0.002361 (  0.003161)
=> nil
```

Reply:
It depends – if we cache each row in the table based on the Release, and make sure to `touch` the release whenever it's deployed, then an _n + 1_ makes sense, as the common case will be that almost all of the releases will be cached and therefore won't trigger any queries. That's preferable to naively preloading everything.

I'd like to add that caching in a separate PR, though, unless it's a blocker for this one.

Reply:
something like this ?
```
deploy = stage.last_deploy
if deploy.success?
  'success'
elsif deploy.running?
  'warning'
else
  'default'
end
```

Reply:
Not quite – I'll push an updated version.